### DISCUSSION TITLE
LFX Mentorship 2023 01-Mar-May Challenge - for #2231

### DISCUSSION THREAD

**Discussion by juntao** [WasmEdge/WasmEdge#2232] [Show and tell]
## Introduction

For those who want to apply for LFX mentorship for #2231, this is a pretest for the application. The goal of this LFX mentorship is to add a feature to the [Fluvio](https://www.fluvio.io/) project so that it can use WasmEdge as an alternative to wasmtime for streaming data processing. We designed this challenge to evaluate the candidates.

## Requirements

1 Create a `filter` wasm module [using Fluvio tools](https://www.fluvio.io/smartmodules/transform/filter/).

2 Study the [`transforms/filter.rs`](https://github.com/infinyon/fluvio/blob/master/crates/fluvio-smartengine/src/transforms/filter.rs) in the [fluvio-smartengine](https://github.com/infinyon/fluvio/tree/master/crates/fluvio-smartengine) crate, and see how it calls the `filter()` function in the Wasm module.

3 Study how to accomplish the same with either the [wasmedge-sdk](https://wasmedge.github.io/WasmEdge/wasmedge_sdk/) or [wasmedge-sys](https://wasmedge.github.io/WasmEdge/wasmedge_sys/) crates. Note: you are not allowed to make any changes to the compiled wasm module from step 1.

4 Modify the `filter.rs` file to replace wasmtime with WasmEdge API calls, and pass Rust compiler check.

## Resources

* [WasmEdge Rust SDK](https://wasmedge.org/book/en/sdk/rust.html)
    * [wasmedge-sdk](https://wasmedge.github.io/WasmEdge/wasmedge_sdk/) is the high level API
    * [wasmedge-sys](https://wasmedge.github.io/WasmEdge/wasmedge_sys/) is the low level API
* [WasmEdge Rust SDK examples](https://github.com/second-state/wasmedge-rustsdk-examples)

## Deadline

Please refer to [the timeline of LFX mentorship 2023 01-Mar-May](https://github.com/cncf/mentoring/tree/main/lfx-mentorship/2023/01-Mar-May).
Since we have to process the internal review and decide, the final date for valid submissions will be February 20, 5:00 PM PDT.

**Comment by Hrushi20** [comment]
@juntao, I'm currently working on this issue. Do I remove the wasmtime API's in fluvio-smartengine crate or do I use `cfg` to support wasmedge along with wasmtime?

**Reply by juntao** [reply]
Eventually, our goal is to support both runtimes through a crate feature flag.

But for the pre-test, you can just swap out wasmtime to demonstrate how to use the WasmEdge SDK. Thanks.

**Reply by Hrushi20** [reply]
Hey @juntao, The `Memory` from `wasmedge-sdk` doesn't support the copy/clone trait. I'm not able to fetch the value inside a MutexLock as I am not able to clone it. Any suggestions? The `fluvio-smartengine` crate use Mutex locks to write data to memory.

**Reply by apepkuss** [reply]
@Hrushi20 Could you please share with us the relevant code piece on your side or a link to it? Thanks!

**Comment by Hrushi20** [comment]
Thanks for the reply. [Code](https://gist.github.com/Hrushi20/73a3428d10d134d877898ca6bbbec20b) regarding the error I'm facing. 

The get method in the recordsCallback can't return the value inside MutexGuard. To return the value I need to clone the MutexGuard. Memory from `wasmedge-sdk` lacks the `clone/copy` trait.

![Screenshot from 2023-01-31 11-05-09](https://user-images.githubusercontent.com/65367566/215675136-98d77242-5fb8-4803-b859-4ed03ae0e29c.png)

![Screenshot from 2023-01-31 11-04-50](https://user-images.githubusercontent.com/65367566/215675165-c6db0516-65ad-4081-ae94-6e03aec2a35b.png)

The RecordsMemory contains pointer to the ouput of the filter. 
Is there a way to return the `Option<RecordsCallback>` from the `get` method without the copy trait???

**Reply by apepkuss** [reply]
Hey @Hrushi20, 

I noticed that some places in the [Code](https://gist.github.com/Hrushi20/73a3428d10d134d877898ca6bbbec20b) you provided can be improved:

-  The first place is that you use the `Memory` type defined in `wasmedge-sys` crate, which is a low-level type, in [line 201](https://gist.github.com/Hrushi20/73a3428d10d134d877898ca6bbbec20b#file-instance-rs-L201). The correct type you should use in the definition of `RecordsMemory` struct is the `Memory` type defined in `wasmedge-sdk` crate. 

- The second place that is worth being mentioned is that [line 103-113](https://gist.github.com/Hrushi20/73a3428d10d134d877898ca6bbbec20b#file-instance-rs-L103-L113). For discussion, I copy and paste the code below, and add a comment in code:

  ```rust
  let copy_records_fn = move |_caller: CallingFrame, inputs: Vec<WasmValue>| -> Result<Vec<WasmValue>, HostFuncError> {
      // retrieve low-level Memory instance here
      let memory = _caller.memory_mut(0).unwrap();
  
      ...
  };
  ```
  To retrieve a high-level `Memory` from a `CallingFrame`, the appropriate way should be
  
  
  ```rust
  use wasmedge_sdk::Caller;
  
  let copy_records_fn = move |_caller: CallingFrame, inputs: Vec<WasmValue>| -> Result<Vec<WasmValue>, HostFuncError> {
      // convert a low-level CallingFrame instance into a high-level Caller instance
      let caller = Caller::new(_caller);
  
      // instead of low-level Memory instance, retrieve high-level Memory instance
      let memory = caller.memory(0).unwrap();
  
      ...
  };
  ```
  The code line of the conversion can guarantee that you always work with high-level objects, instead of low-level ones.

For the issue of implementing the `Copy` and/or `Clone` trait for `Memory`, I believe we can provide support. Before doing it, however, we have to consider the performance loss. In essence, the low-level `Memory` type defined in `wasmedge-sys` is a wrapper of a raw pointer of  `*mut WasmEdge_MemoryInstanceContext` type. When we wanna drop such a `Memory` instance, we have to call the `WasmEdge_MemoryInstanceDelete` function manually. To avoid dangling pointers, we need to use `Arc` to wrap it so that the drop operation can be performed automatically and correctly. I am not very sure if the introduction of `Arc` could cause any performance loss. So please let us check the idea first, and we'll give a solution to the issue asap. Our goal is to implement both `Copy` and `Clone` traits for both high-level and low-level `Memory` types. But it is strongly recommended to use high-level type in this project. Thanks a lot!

**Reply by Hrushi20** [reply]
Thanks @apepkuss, I didn't know we could convert the caller to high-level. I've updated the code. Do I just submit the code with this pending task or wait for `Copy/Clone` trait to be added?

**Reply by apepkuss** [reply]
Would it be blocking anything, like test cases or other things if submit the code without the support for the traits? We'll provide support asap, but still need some time to do it. Thanks!

**Reply by Hrushi20** [reply]
It would block the test cases. The `Clone/Copy` trait gets the output of the function from web assembly.

**Reply by apepkuss** [reply]
PR #2246 has been created and is waiting for review.

**Reply by Hrushi20** [reply]
I've noticed that the PR has been merged. How do I use the latest wasmedge-sdk crate?

**Reply by apepkuss** [reply]
For this round of release, we'll publish a new release of `wasmedge-sdk` crate when the new version of the runtime is released. It means we will not publish a new version of `wasmedge-sdk` with the changes in PR #2246 right now. But you still can use it in this project by replacing [`wasmedge-sdk` dependecy in Cargo.toml](https://github.com/Hrushi20/fluvio-smart-engine-wasmedge/blob/main/Cargo.toml#L25) with the line below:

```toml
wasmedge-sdk = { git = "https://github.com/WasmEdge/WasmEdge", branch = "master" }
```
Try it on your side, and let me know if any further questions. Thanks!

**Comment by Hrushi20** [comment]
Also another challenge which I'm facing is writing data to webassembly linear memory. 
The Wasmtime provides a method `memory.data_ptr(store).offset(guest_ptr_offset);` which returns the pointer to where the data needs to be written. The guest_ptr_offset is added to the base pointer which is then returned. 

``` rust

// Wasmtime 
  unsafe {
        let raw = memory.data_ptr(store).offset(guest_ptr_offset);
        println!("Data written at: {:?}",raw as i32);
        raw.copy_from(bytes.as_ptr(), bytes.len());                           
    }
```

However, in wasmedge, we don't have a method that returns the base pointer, in the host's address space. 

``` rust

   // Wasmedge
    memory.write(bytes, guest_ptr_offset as u32).expect("Coudlnt' write data to memory");
```

Here the data is being written to the guest_ptr_offset address. Due to this I'm getting invalid memory reference while calling the wasm function with the required inputs.

[Gist](https://gist.github.com/Hrushi20/b1b5834592159ee5d69d309e2cb05434#file-memory-rs). 



WasmEdge output- (Error: invalid memory reference)

![Screenshot from 2023-01-31 23-49-48](https://user-images.githubusercontent.com/65367566/215848820-5b76a5c1-042b-4d1c-bb41-5b07e7b032ce.png)

Wasmtime output- 

![Screenshot from 2023-01-31 23-49-58](https://user-images.githubusercontent.com/65367566/215848828-dbeccaa5-9d1e-42ed-8642-7e97ff7cc046.png)

**Reply by apepkuss** [reply]
Thanks for your report. 

As for how to use the `write` method of `Memory` correctly, I extract the code snippet from [the test case for memory](https://github.com/WasmEdge/WasmEdge/blob/master/bindings/rust/wasmedge-sdk/src/externals/memory.rs#L261-L271) and copy and paste it below as a reference. 

```rust
...
// write data
let data = vec![1; 10];
let result = memory.write(data.as_slice(), 10);
assert!(result.is_ok());
// read data after the write operation 
let result = memory.read(10, 10);
assert!(result.is_ok());
let data = result.unwrap();
assert_eq!(data, vec![1; 10]);
...
```
Back to your code, I cannot tell the stem reason for the failure. The invocation of the `write` method seems correct. Could you please share with us the link to the full code? We'll check it again. Thanks a lot!

**Comment by Hrushi20** [comment]
I've tried implementing the required feature. The link to the repository is https://github.com/Hrushi20/fluvio-smart-engine-wasmedge.
Facing minor errors which can be resolved :smile:

**Reply by juntao** [reply]
Thank you! Really appreciate the suggestions you made through this effort.

**Comment by Hrushi20** [comment]
I ran the wasm function by coding it in the [main.rs](https://github.com/Hrushi20/fluvio-smart-engine-wasmedge/blob/main/src/main.rs) file and it works as expected. I'm beginner at rust so I guess I kinda messed up the code due to which the tests are failing (invalid memory reference). Hence I just coded the main functionality in the main.rs file and things work as expected. I'll surely learn and get better at rust. Thank You :smile:

**Reply by Hrushi20** [reply]
I have also added the statistics to the pretest by configuring it and passing it to the executor.

**Reply by Hrushi20** [reply]
Added tests to the main.rs file.

**Comment by xxchan** [comment]
Hi, I'm also interested in this project and I submitted the application at LFX. Here's my pretest: https://github.com/xxchan/fluvio/pull/1
